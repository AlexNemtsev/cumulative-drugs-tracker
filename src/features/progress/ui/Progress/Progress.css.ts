import { style } from '@vanilla-extract/css';

import { container } from '@/shared/styles/container.css';

export const progressBar = style({
  height: 25,
});

export const card = style([
  container,
  {
    padding: 16,
    vars: {
      '--card-border-width': '0',
    },
  },
]);
