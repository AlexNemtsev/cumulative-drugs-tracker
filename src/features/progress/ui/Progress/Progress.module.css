.progressBar {
  height: 25px;
}

.card {
  --card-border-width: 0;

  padding: 16px;
  border-radius: 30px;
  background-color: #fff;
}