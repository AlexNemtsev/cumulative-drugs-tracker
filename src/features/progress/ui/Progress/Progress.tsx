import { Flex, Text, Card, Progress as ProgressBar } from '@radix-ui/themes';
import { useState } from 'react';

import { RecordModal } from '@/entities/record/ui/RecordModal';
import { AppSettings } from '@/shared/appSettings';
import { useRecords } from '@/shared/providers/RecordsProvider';
import type { RecordType } from '@/shared/types/Record';
import { AddButton } from '@/shared/ui/AddButton';

import { card, progressBar } from './Progress.css';

const dateConfig: Intl.DateTimeFormatOptions = {
  day: 'numeric',
  month: 'long',
  year: 'numeric',
};

export const Progress = () => {
  const { records, addRecord } = useRecords();
  const [isRecordModalOpened, setIsRecordModalOpened] = useState(false);

  const openRecordModal = () => {
    setIsRecordModalOpened(true);
  };

  const closeRecordModal = () => {
    setIsRecordModalOpened(false);
  };

  const handleAddRecord = async (record: RecordType) => {
    await addRecord(record);

    closeRecordModal();
  };

  const cumulativeDose = records.reduce<number>((acc, record) => {
    const dose = +record.dose;

    return acc + dose;
  }, 0);

  const progressBarValue = (cumulativeDose / AppSettings.TARGET_DOSE) * 100;

  const doseRemained = AppSettings.TARGET_DOSE - cumulativeDose;
  const etaDays = doseRemained / +AppSettings.DAY_TARGET;

  const expectedDay = new Date();
  expectedDay.setDate(expectedDay.getDate() + Math.ceil(etaDays));

  return (
    <Flex direction="column" gap="5">
      <Card className={card}>
        <Flex direction="column" gap="4">
          <Text size="6">Суммарная доза</Text>
          <ProgressBar value={progressBarValue} className={progressBar} />
          <Text size="6">
            Принято {cumulativeDose} мг из {AppSettings.TARGET_DOSE} мг
          </Text>
          <Text size="5">
            Ожидаемое завершение – {expectedDay.toLocaleDateString('ru-RU', dateConfig)}
          </Text>
        </Flex>
      </Card>
      <AddButton onClick={openRecordModal} />
      <RecordModal
        isOpen={isRecordModalOpened}
        onSubmit={handleAddRecord}
        onCancel={closeRecordModal}
      />
    </Flex>
  );
};
