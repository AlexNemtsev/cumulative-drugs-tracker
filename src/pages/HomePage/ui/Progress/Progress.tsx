import { Flex, Text, Card, Progress as ProgressBar } from '@radix-ui/themes';

import { card, progressBar } from './Progress.css';

type Props = {
  takenDose: number;
  totalTargetDose: number;
  dayTargetDose: number;
  currentDate: Date;
};

const dateConfig: Intl.DateTimeFormatOptions = {
  day: 'numeric',
  month: 'long',
  year: 'numeric',
};

export const Progress = (props: Props) => {
  const { dayTargetDose, takenDose, totalTargetDose, currentDate } = props;

  const progressBarValue = (takenDose / totalTargetDose) * 100;

  const doseRemained = totalTargetDose - takenDose;
  const etaDays = doseRemained / dayTargetDose;

  currentDate.setDate(currentDate.getDate() + Math.ceil(etaDays));

  return (
    <Card className={card} data-testid="progress">
      <Flex direction="column" gap="4">
        <Text size="6">Суммарная доза</Text>
        <ProgressBar value={progressBarValue} className={progressBar} />
        <Text size="6">
          Принято {takenDose} мг из {totalTargetDose} мг
        </Text>
        {progressBarValue < 100 ? (
          <Text size="5">
            Ожидаемое завершение – {currentDate.toLocaleDateString('ru-RU', dateConfig)}
          </Text>
        ) : (
          <Text>Приём лекарства завершён</Text>
        )}
      </Flex>
    </Card>
  );
};
