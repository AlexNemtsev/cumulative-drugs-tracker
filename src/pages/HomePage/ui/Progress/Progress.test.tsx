import { render, screen } from '@testing-library/react';

import { Progress } from './Progress';

const fixedData = {
  dayTargetDose: 20,
  totalTargetDose: 100,
  currentDate: new Date('2025-07-11'),
};

test.each([
  { ...fixedData, takenDose: 0 },
  { ...fixedData, takenDose: 40 },
  { ...fixedData, takenDose: 100 },
])(
  'Компонент Progress должен отрисоваться с принятой дозировкой $takenDose',
  ({ currentDate, dayTargetDose, takenDose, totalTargetDose }) => {
    render(
      <Progress
        currentDate={currentDate}
        dayTargetDose={dayTargetDose}
        takenDose={takenDose}
        totalTargetDose={totalTargetDose}
      />
    );

    const progressContainer = screen.getByTestId('progress');

    expect(progressContainer).toMatchSnapshot();
  }
);
