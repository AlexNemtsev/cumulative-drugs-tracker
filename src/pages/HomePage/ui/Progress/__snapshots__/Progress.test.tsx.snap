// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Компонент Progress должен отрисоваться с принятой дозировкой +0 1`] = `
<div
  class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface Progress_card__10lthzj1 container_container__1h2gq2x0"
  data-testid="progress"
>
  <div
    class="rt-Flex rt-r-fd-column rt-r-gap-4"
  >
    <span
      class="rt-Text rt-r-size-6"
    >
      Суммарная доза
    </span>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      aria-valuetext="0%"
      class="rt-ProgressRoot rt-r-size-2 rt-variant-surface Progress_progressBar__10lthzj0"
      data-max="100"
      data-state="loading"
      data-value="0"
      role="progressbar"
      style="--progress-value: 0;"
    >
      <div
        class="rt-ProgressIndicator"
        data-max="100"
        data-state="loading"
        data-value="0"
      />
    </div>
    <span
      class="rt-Text rt-r-size-6"
    >
      Принято 
      0
       мг из 
      100
       мг
    </span>
    <span
      class="rt-Text rt-r-size-5"
    >
      Ожидаемое завершение – 
      16 июля 2025 г.
    </span>
  </div>
</div>
`;

exports[`Компонент Progress должен отрисоваться с принятой дозировкой 40 1`] = `
<div
  class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface Progress_card__10lthzj1 container_container__1h2gq2x0"
  data-testid="progress"
>
  <div
    class="rt-Flex rt-r-fd-column rt-r-gap-4"
  >
    <span
      class="rt-Text rt-r-size-6"
    >
      Суммарная доза
    </span>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="40"
      aria-valuetext="40%"
      class="rt-ProgressRoot rt-r-size-2 rt-variant-surface Progress_progressBar__10lthzj0"
      data-max="100"
      data-state="loading"
      data-value="40"
      role="progressbar"
      style="--progress-value: 40;"
    >
      <div
        class="rt-ProgressIndicator"
        data-max="100"
        data-state="loading"
        data-value="40"
      />
    </div>
    <span
      class="rt-Text rt-r-size-6"
    >
      Принято 
      40
       мг из 
      100
       мг
    </span>
    <span
      class="rt-Text rt-r-size-5"
    >
      Ожидаемое завершение – 
      19 июля 2025 г.
    </span>
  </div>
</div>
`;

exports[`Компонент Progress должен отрисоваться с принятой дозировкой 100 1`] = `
<div
  class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface Progress_card__10lthzj1 container_container__1h2gq2x0"
  data-testid="progress"
>
  <div
    class="rt-Flex rt-r-fd-column rt-r-gap-4"
  >
    <span
      class="rt-Text rt-r-size-6"
    >
      Суммарная доза
    </span>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      aria-valuetext="100%"
      class="rt-ProgressRoot rt-r-size-2 rt-variant-surface Progress_progressBar__10lthzj0"
      data-max="100"
      data-state="complete"
      data-value="100"
      role="progressbar"
      style="--progress-value: 100;"
    >
      <div
        class="rt-ProgressIndicator"
        data-max="100"
        data-state="complete"
        data-value="100"
      />
    </div>
    <span
      class="rt-Text rt-r-size-6"
    >
      Принято 
      100
       мг из 
      100
       мг
    </span>
    <span
      class="rt-Text"
    >
      Приём лекарства завершён
    </span>
  </div>
</div>
`;
